antarest/study/directory_service.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) 2025, RTE (https://www.rte-france.com)
+#
+# See AUTHORS.txt
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# SPDX-License-Identifier: MPL-2.0
+#
+# This file is part of the Antares project.
+import logging
+import uuid
+from typing import List, Sequence
+
+from antarest.core.requests import UserHasNotPermissionError
+from antarest.login.model import Group
+from antarest.study.directory_exceptions import (
+    DirectoryAlreadyExistsError,
+    DirectoryCycleError,
+    DirectoryNotEmptyError,
+    DirectoryNotFoundError,
+    DirectoryPermissionError,
+)
+from antarest.study.model import (
+    Directory,
+    DirectoryCreation,
+    DirectoryMetadata,
+    DirectoryUpdate,
+)
+from antarest.study.repository import AccessPermissions, DirectoryRepository, StudyMetadataRepository
+from antarest.study.service import StudyService
+
+logger = logging.getLogger(__name__)
+
+
+class DirectoryService:
+    """Service for managing directories."""
+
+    def __init__(
+        self,
+        directory_repository: DirectoryRepository,
+        study_repository: StudyMetadataRepository,
+        study_service: StudyService,
+    ):
+        self.directory_repository = directory_repository
+        self.study_repository = study_repository
+        self.study_service = study_service
+
+    def list_directories(self) -> List[DirectoryMetadata]:
+        """List all directories the user has access to."""
+        access_permissions = AccessPermissions.for_current_user()
+        directories = self.directory_repository.get_all(access_permissions)
+        return [directory.to_metadata() for directory in directories]
+
+    def create_directory(
+        self,
+        data: DirectoryCreation,
+        owner_id: int,
+        group_ids: Sequence[str],
+    ) -> DirectoryMetadata:
+        access_permissions = AccessPermissions.for_current_user()
+        if data.parent_id:
+            parent = self.directory_repository.get(data.parent_id)
+            if parent is None:
+                raise DirectoryNotFoundError(data.parent_id)
+
+            if not self.directory_repository.has_permission(parent, access_permissions):
+                raise DirectoryPermissionError("You don't have permission to create directories in this parent")
+
+        if self.directory_repository.has_duplicate_name(data.name, data.parent_id):
+            raise DirectoryAlreadyExistsError(data.name)
+
+        directory = Directory(
+            id=str(uuid.uuid4()),
+            name=data.name,
+            parent_id=data.parent_id,
+            owner_id=owner_id,
+        )
+
+        session = self.directory_repository.session
```

Comment:
we'd better use the group repository here, the repositories are supposed to hide the session as an implementation detail